src/indexed_crate.rs

```diff
@@ -41,6 +41,166 @@ pub struct IndexedCrate<'a> {
 }
 
+/// Map a Key to a List (Vec) of values
+///
+/// It also has some nice operations for pushing a value to the list, or extending the list with
+/// many values.
+struct MapList<K, V>(HashMap<K, Vec<V>>);
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> FromIterator<(K, V)> for MapList<K, V> {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let iter = iter.into_iter();
+        let (min, max) = iter.size_hint();
+        // Reserving this space doesn't give a measurable gain, but neither does it hurt so we are
+        // keeping it around
+        let mut map = Self::with_capacity(max.unwrap_or(min));
+        for (key, value) in iter {
+            map.insert(key, value);
+        }
+        map
+    }
+}
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> Extend<(K, V)> for MapList<K, V> {
+    #[inline]
+    fn extend<T: IntoIterator<Item = (K, V)>>(&mut self, iter: T) {
+        // We could use Iterator::size_hint here to reserve some space, but I measured a 2%-3%
+        // regression when doing that.
+        for (key, value) in iter.into_iter() {
+            self.insert(key, value);
+        }
+    }
+}
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> MapList<K, V> {
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(HashMap::with_capacity(capacity))
+    }
+
+    #[inline]
+    pub fn into_inner(self) -> HashMap<K, Vec<V>> {
+        self.0
+    }
+
+    #[inline]
+    pub fn insert(&mut self, key: K, value: V) {
+        match self.0.entry(key) {
+            Entry::Occupied(mut entry) => entry.get_mut().push(value),
+            Entry::Vacant(entry) => {
+                entry.insert(vec![value]);
+            }
+        }
+    }
+}
+
+/// Build the imports index
+///
+/// When compiled using the `rayon` feature, build it in parallel. Specifically, this paralelizes
+/// getting the [`IndexedCrate::publicly_importable_names`] over the items in the index.
+fn build_imports_index<'a>(
+    index: &'a HashMap<Id, Item>,
+    value: &IndexedCrate<'a>,
+) -> MapList<Path<'a>, (&'a Item, Modifiers)> {
+    let iter = index.iter();
+
+    iter.filter_map(|(_id, item)| {
+        if !supported_item_kind(item) {
+            return None;
+        }
+        let importable_paths = value.publicly_importable_names(&item.id);
+        let iter = importable_paths.into_iter();
+        Some(
+            iter.map(move |importable_path| {
+                (importable_path.path, (item, importable_path.modifiers))
+            }),
+        )
+    })
+    .flatten()
+    .collect()
+}
+
+/// Build the impl index
+///
+/// When compiled using the `rayon` feature, build it in parallel. Specifically, this paralelizes
+/// the work of gathering all of the impls for the items in the index.
+fn build_impl_index(index: &HashMap<Id, Item>) -> MapList<ImplEntry<'_>, (&Item, &Item)> {
+    let iter = index.iter();
+    iter.filter_map(|(id, item)| {
```

Comment:
I mentioned in a parallel PR that as a personal preference thing, I find code easier to read if it creates and immediately consumes iterators instead of going through an intermediate variable.

If you don't mind, I'd appreciate it if you could tweak the code here and elsewhere in this PR in that manner, to help me maintain this code with less work in the long run:
```suggestion
    index.iter().filter_map(|(id, item)| {
```

Reply:
We introduce the variable back once we parallelize it with rayon:

```rust
#[cfg(feature = "rayon")]
let iter = iter.par_iter();
#[cfg(not(feature = "rayon"))]
let iter = iter.iter();
```

I also prefer it the other way around.

I can switch it around as then reverting the rayon feature would leave the code in a better state.

Reply:
Ah that explains it, thank you.

I agree, adding the variable in the `rayon`-specific PR would be ideal.

Reply:
Fixed!